src/charm.py

```diff
@@ -135,11 +153,31 @@ def reconcile(self, event: EventBase) -> None:
         if getattr(event, "departing_unit", None) == self.unit:
             return
 
-        outdated_status = []
-        # Maintain the correct app status
-        if self.unit.is_leader():
-            if self.state.opensearch_server:
-                outdated_status.append(MSG_DB_MISSING)
+        # 2. Restart on config change
+        if (
+            self.config_manager.config_changed()
+            and self.state.unit_server.started
+            and self.upgrade_events.idle
+        ):
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
+            # No point in setting any status -- would be wiped out by rollingops after the restert
+            return
+
+        # 3. Maintain the correct app status
+        # No further actions below but only status settings
+
+        # Block until Opensearch is available and it's a compatible version
+        if self.state.opensearch_server:
+            outdated_status.append(MSG_STATUS_DB_MISSING)
+        else:
+            set_global_status(self, BlockedStatus(MSG_STATUS_DB_MISSING))
+            return
```

Comment:
I was thinking about this myself, and at the end consciously decided not to go for such.

_"Explicit is better than implicit"_ 

I prefer to reduce the level of redirections an keep the code clear and straightforward. I agree that this this pattern is repetitive, however I still rather prefer it repetitive but clear rather than "yet another" "magical helpler function" that is calling other helper functions, and  the reader needs to seek and resolve this whole "tangled-up" workflow in other functions/files. 

Since there is no issue on readability or clarity regarding the current solution, (the only issue is that it's "too verbose/repetitive") I'd prefer to leave it as such -- at least until I may find a way I'm also feeling comofortable with.